src/instrument.ts

```diff
@@ -246,3 +250,284 @@ class noopSpan implements Span {
     return this;
   }
 }
+
+interface TraceWrapped extends Function {
+  traceWrapped?: Function;
+}
+
+function performWrap(
+  klass: Function,
+  spanName: string,
+  original: TraceWrapped
+) {
+  /*
+  original would be for example:
+    const [rows] = await database.run('SELECT 1');
+  or
+    database.run('SELECT 1', (err, rows) => {
+        ...
+    });
+  */
+  if (original.traceWrapped) {
+    return original;
+  }
+
+  const traced: TraceWrapped = function (
+    this: any,
+    ...args: any[]
+  ):
+    | void
+    | Promise<unknown>
+    | PartialResultStream
+    | EventEmitter
+    | AsyncIterable<unknown> {
+    const lastArg = args[args.length - 1];
+    const hasCallback = typeof lastArg === 'function';
+    const firstArg = args[0];
+
+    const opts: traceConfig = {opts: this.observabilityOptions || {}};
+
+    if (typeof firstArg === 'string' || (firstArg as SQLStatement)) {
+      opts.sql = firstArg;
+    }
+
+    return startTrace(spanName, opts, span => {
+      // Case 1. We have a callback function, plainly wrap it
+      // with a passthrough function that starts the span on invocation
+      // and then ends it once results are returned from the original.
+      if (hasCallback) {
+        const callback = lastArg;
+        const wrappedFn = function (...args: any) {
+          const errIndex = Array.from(args).findIndex(
+            arg => arg instanceof Error
+          );
+          const err: Error = errIndex !== -1 ? args[errIndex] : null;
+          setSpanError(span, err);
+          span.end();
+          return callback(...args);
+        };
+
+        // Let's copy the name of the wrapping callback
+        // to be that of the original callback function.
+        Object.defineProperty(wrappedFn, 'name', {value: callback.name});
+        Object.defineProperty(wrappedFn, '__original_callback', {
+          value: callback,
+        });
+        args[args.length - 1] = wrappedFn;
+
+        // Finally invoke the function with its modified callback argument.
+        return original.apply(this, args);
+      }
+
+      let originalResult: any = null;
+
+      try {
+        originalResult = original.apply(this, args);
+      } catch (err: any) {
```

Comment:
What is error events are emitted instead of throwing error , how will those cases be handled ?